src/common/src/license/manager.rs

```diff
@@ -0,0 +1,288 @@
+    Unknown(String),
+}
+
+/// The content of a license.
+///
+/// We use JSON Web Token (JWT) to represent the license. This struct is the payload.
+// TODO: Shall we add a version field?
+#[derive(Debug, Clone, Deserialize)]
+#[serde(rename_all = "snake_case")]
+pub(super) struct License {
+    /// Subject of the license.
+    ///
+    /// See <https://tools.ietf.org/html/rfc7519#section-4.1.2>.
+    #[allow(dead_code)]
+    pub sub: String,
+
+    /// Issuer of the license.
+    ///
+    /// See <https://tools.ietf.org/html/rfc7519#section-4.1.1>.
+    #[allow(dead_code)]
+    pub iss: Issuer,
+
+    /// Tier of the license.
+    pub tier: Tier,
+
+    /// Expiration time in seconds since UNIX epoch.
+    ///
+    /// See <https://tools.ietf.org/html/rfc7519#section-4.1.4>.
+    pub exp: u64,
+}
+
+impl Default for License {
+    /// The default license is a free license that never expires.
+    ///
+    /// Used when `license_key` is unset or invalid.
+    fn default() -> Self {
+        Self {
+            sub: "default".to_owned(),
+            tier: Tier::Free,
+            iss: Issuer::Prod,
+            exp: u64::MAX,
+        }
+    }
+}
+
+/// The error type for invalid license key when verifying as JWT.
+#[derive(Debug, Clone, Error)]
+#[error("invalid license key")]
+pub struct LicenseKeyError(#[source] jsonwebtoken::errors::Error);
+
+struct Inner {
+    license: Result<License, LicenseKeyError>,
+}
+
+/// The singleton license manager.
+pub(crate) struct LicenseManager {
+    inner: RwLock<Inner>,
+}
+
+static PUBLIC_KEY: LazyLock<DecodingKey> = LazyLock::new(|| {
+    DecodingKey::from_rsa_pem(include_bytes!("key.pub"))
+        .expect("invalid public key for license validation")
+});
+
+impl LicenseManager {
+    /// Create a new license manager with the default license.
+    fn new() -> Self {
+        Self {
+            inner: RwLock::new(Inner {
+                license: Ok(License::default()),
+            }),
+        }
+    }
+
+    /// Get the singleton instance of the license manager.
+    pub fn get() -> &'static Self {
+        static INSTANCE: LazyLock<LicenseManager> = LazyLock::new(LicenseManager::new);
+        &INSTANCE
+    }
+
+    /// Refresh the license with the given license key.
+    pub fn refresh(&self, license_key: &str) {
+        let mut inner = self.inner.write().unwrap();
+
+        // Empty license key means unset. Use the default one here.
+        if license_key.is_empty() {
+            inner.license = Ok(License::default());
+            return;
+        }
+
+        // TODO: shall we also validate `nbf`(Not Before)?
+        let mut validation = Validation::new(Algorithm::RS512);
+        // Only accept `prod` issuer in production, so that we can use license keys issued by
+        // the `test` issuer in development without leaking them to production.
+        validation.set_issuer(&[
+            "prod.risingwave.com",
+            #[cfg(debug_assertions)]
+            "test.risingwave.com",
```

Comment:
So `test.risingwave.com` only works for debug builds. IIUC, the tests using release build (such as the `main-cron` and benchmarks) will need to set up a `prod.risingwave.com` license to run, right?

Reply:
You're right. We should
- either generate a `prod` key and set them for benchmarks (with potential risk of leaking it)
- or use a better compiler flag to distinguish between test/development build and production build, which we don't have now AFAIK

Do you have any preference or other ideas? Also cc @xxchan

BTW, `ci-release` profile is used in `main-cron`, which enables debug assertions.

Reply:
Also related https://github.com/risingwavelabs/risingwave/pull/17456#issuecomment-2190671517

Reply:
> or use a better compiler flag to distinguish between test/development build and production build, which we don't have now AFAIK

You reminds me of this: https://github.com/tikv/tikv/blob/master/etc/cargo.config.dist  "really prod/dist config" (e.g., fat LTO) vs "release/bench" config. I also wanted a "release without LTO" before, because LTO compiles for too long. 🤣

Anyway, I feel generate a `prod` key for benchmark is enough. And we can manage it in test env like AWS secrets.